lighthouse-core/gather/fetcher.js

```diff
@@ -84,19 +88,105 @@ class Fetcher {
   }
 
   /**
-   * Requires that `driver.enableRequestInterception` has been called.
+   * Requires that `fetcher.enable` has been called.
    *
    * Fetches any resource in a way that circumvents CORS.
    *
    * @param {string} url
-   * @param {{timeout: number}} options timeout is in ms
+   * @param {{timeout: number}=} options timeout is in ms
    * @return {Promise<string>}
    */
-  async fetchResource(url, {timeout = 500}) {
+  async fetchResource(url, options = {timeout: 500}) {
     if (!this._enabled) {
-      throw new Error('Must call `enableRequestInterception` before using fetchResource');
+      throw new Error('Must call `enable` before using fetchResource');
+    }
+
+    const milestone = await this.driver.getBrowserVersion().then(v => v.milestone);
+    if (milestone >= 88) {
+      return await this._fetchResourceOverProtocol(url, options);
+    }
+    return await this._fetchResourceIframe(url, options);
+  }
+
+  /**
+   * @param {string} handle
+   * @param {{timeout: number}=} options,
+   * @return {Promise<string>}
+   */
+  async _readIOStream(handle, options = {timeout: 500}) {
+    const startTime = Date.now();
+
+    let ioResponse;
+    let data = '';
+    while (!ioResponse || !ioResponse.eof) {
+      const elapsedTime = Date.now() - startTime;
+      if (elapsedTime > options.timeout) {
+        throw new Error('Waiting for the end of the IO stream exceeded the allotted time.');
+      }
+      ioResponse = await this.driver.sendCommand('IO.read', {handle});
+      const responseData = ioResponse.base64Encoded ?
+        Buffer.from(ioResponse.data, 'base64').toString('utf-8') :
+        ioResponse.data;
+      data = data.concat(responseData);
+    }
+
+    return data;
+  }
+
+  /**
+   * @param {string} url
+   */
+  async _fetchStream(url) {
+    await this.driver.sendCommand('Network.enable');
+    const frameTreeResponse = await this.driver.sendCommand('Page.getFrameTree');
+    const networkResponse = await this.driver.sendCommand('Network.loadNetworkResource', {
+      frameId: frameTreeResponse.frameTree.frame.id,
+      url,
+      options: {
+        disableCache: true,
+        includeCredentials: true,
+      },
+    });
+    await this.driver.sendCommand('Network.disable');
+
+    if (!networkResponse.resource.success || !networkResponse.resource.stream) {
+      const statusCode = networkResponse.resource.httpStatusCode || '';
+      throw new Error(`Loading network resource failed (${statusCode})`);
     }
 
+    return networkResponse.resource.stream;
+  }
+
+  /**
+   * @param {string} url
+   * @param {{timeout: number}=} options timeout is in ms
+   * @return {Promise<string>}
+   */
+  async _fetchResourceOverProtocol(url, options = {timeout: 500}) {
```

Comment:
since these are internal and `fetchResource()` sets default values, would it work to make the `options` and `options.timeout` required for `_fetchResourceOverProtocol` and `_readIOStream`? It's harder to reason about where the timeout is being set when down a few levels even though in practice it will always be set by the beginning of `fetchResource()`.